src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -605,6 +731,107 @@ public void withLock(boolean waitForRecovery, ChannelConsumer consumer) throws I
+            try {
+                Channel tempChannel = getChannel(true);
+                while (true) {
+                    try {
+                        consumer.consume(tempChannel);
+                        break;
+                    } catch (IOException | ShutdownSignalException e) {
+                        var currentValue = iterator.next();
+                        int recoveryDelay = currentValue.getDelay();
+                        LOGGER.warn("Retrying publishing #{}, waiting for {}ms. Reason: {}", currentValue.getTryNumber(), recoveryDelay, e);
+                        TimeUnit.MILLISECONDS.sleep(recoveryDelay);
+
+                        // We should not recover the channel if its connection is closed
+                        // If we do that the channel will be also auto recovered by RabbitMQ client
+                        // during connection recovery, and we will get two new channels instead of one closed.
+                        if (!tempChannel.isOpen() && tempChannel.getConnection().isOpen()) {
+                            tempChannel = recreateChannel();
+                        }
+                    }
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        public <T> T retryingConsumeWithLock(ChannelMapper<T> mapper, ConnectionManagerConfiguration configuration) throws InterruptedException, IOException {
+            Iterator<RetryingDelay> iterator = null;
+            IOException exception;
+            Channel tempChannel = null;
+            boolean isChannelClosed = false;
+            while (true) {
+                lock.lock();
+                try {
+                    if (tempChannel == null) {
+                        tempChannel = getChannel();
+                    } else if (isChannelClosed) {
+                        tempChannel = recreateChannel();
+                    }
+
+                    var tag = mapper.map(tempChannel);
+                    isSubscribed = true;
+                    return tag;
+                } catch (IOException e) {
+                    var reason = tempChannel.getCloseReason();
+                    isChannelClosed = reason != null;
+                    if (isChannelClosed && reason.getMessage().contains("reply-text=RESOURCE_LOCKED")) {
+                        throw e;
+                    }
+                    exception = e;
+                } finally {
+                    lock.unlock();
+                }
+                iterator = handleAndSleep(configuration, iterator, "Retrying consume", exception);
+            }
+        }
+
+        public Optional<SubscriptionCallbacks> getCallbacksForRecovery(Channel channelToRecover) {
+            lock.lock();
+            SubscriptionCallbacks callbacks = null;
+            try {
+                if (channel != channelToRecover) {
+                    LOGGER.warn("Channel already recovered");
```

Comment:
This case looks like we have bug in our code.
Maybe throw an exception in this case

Reply:
Exception will not be handled because this code executed in executor service.

Reply:
Entry point of our tasks should have try ... catch block to catch all Exceptions.